File: src/benchmark.jl
using DataFrames, TestImages, PyCall, BenchmarkTools, Images
using CSV
function start_benchmark()

    # # Basic Operations
    # io = DataFrame(
    #     operations=["load", "save", "testimage"]timetaken - jl =
    #         ["", "", ""]timetaken - py = ["", "", ""],
    # )
    # ImageMorphology.jl

    println("ImageMorphology Benchmark")
    imagemorphologyjl = DataFrame(
        type=[
            "morphology",
            "morphology",
            "morphology",
            "morphology",
            "morphology",
            "morphology",
        ], 
        operations=[
            # Operations
            "dilate",
            "erode",
            "opening",
            "closing",
            "tophat",
            "bothat",
        ],
        timetakenjl=[0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        resultlink=["/", "/", "/", "/", "/", "/"],
        timetakenpy=[0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        resultlinkpy=["/", "/", "/", "/", "/", "/"],
    )
    count = 1
    original = load("results/original.png")
    # save("results/original.png", original)
    # println(imagemorphologyjl)
    for i in imagemorphologyjl[!, 2]
        benchmark = @benchmark getfield(Images, Symbol($i))($original)
        # print(benchmark)
        result = getfield(Images, Symbol(i))(original)
        save("results/julia/$i.png", result)
        imagemorphologyjl[!, 3][count] = mean(benchmark).time * 10^-12 *10^3
        imagemorphologyjl[!, 4][count] = "results/julia/$i.png"
        count += 1
    end
    println(imagemorphologyjl)
    CSV.write("result.csv", imagemorphologyjl)
    # registration = DataFrame(
    #     operations=["Min", "Max", ",Mean", "Entropy", "PSNR", "SSIM"],
    #     jlway=["minfinite", "maxfinite", "meanfinite", "entropy", "PSNR", "SSIM"],
    #     timetakenjl=["", "", "", "", "", ""],
    #     jlresult=["", "", "", "", "", ""],
    #     pyway=[
    #         "minimum",
    #         "maximum",
    #         "mean",
    #         "entropy",
    #         "compare_psnr",
    #         "compare_ssim",
    #     ],
    #     timetaken - py=["", "", "", "", "", ""],
    #     pyresult=["", "", "", "", "", ""],
    # )
    # println(registration)
    # for i in  registration[!, 1]
    #     benchmark = @benchmark getfield(Images, Symbol($i))($original)
    #     result = getfield(Images, Symbol(i))(original)
    #     save("results/julia/$i.png", mosaicview(original, result))
    #     imagemorphologyjl[!, 2][count] = string(mean(benchmark).time)
    #     imagemorphologyjl[!, 3][count] = "results/julia/$i.png"
    #     count += 1
    # end
end

start_benchmark()

File: src/benchmark.py
import skimage
from skimage import data
from skimage.util import img_as_ubyte
# from skimage.morphology import (erosion, dilation, opening, closing,  # noqa
#                                 white_tophat)
# from skimage.morphology import black_tophat, skeletonize, convex_hull_image  # noqa
from skimage.morphology import disk  # noqa
from skimage import io
import timeit
import time
import pandas as pd

morphology = "morphology"
submodule = getattr(skimage,morphology)

orig_phantom = img_as_ubyte(data.shepp_logan_phantom())

list = ["dilation","erosion","opening","closing","white_tophat","black_tophat"]
data = pd.read_csv("result.csv")
# print(data)
count = 0
for i in list:
    selem = disk(6)
    start_time = timeit.default_timer()
    result =getattr(submodule,i)(orig_phantom, selem)
    timetaken = timeit.default_timer() - start_time
    data.at[count,"timetakenpy"] = timetaken
    resultlink = "results/python/{}.png".format(i)
    data.at[count,"resultlinkpy"] = resultlink
    count +=1
    io.imsave("results/python/{}.png".format(i), result)

print(data)
data.to_csv('result.csv')

data["py/jl"] = data["timetakenpy"]/data["timetakenjl"]
print(data)
data.to_csv('result.csv')